accepted/2022/enhance-vector-codegen.md

```diff
@@ -0,0 +1,679 @@
+
+- When the JIT encounters a `Vectorize.If` intrinsic in a conditional block, the JIT will:
+
+  - Clone the if condition and associated arm with the `Vectorize.If`, creating an arm for each available vector length on the platform. 
+
+  - In each cloned arm any `Vector<T>` operations will be translated to `VectorXX<T>` operations, where XX represents the vector length for that arm.
+
+  - The following example demonstrates the expansion:
+
+    ```C#
+    void unsafe foo(Span<int> sp)
+    {
+      if (Vectorize.If(sp.Length >= Vector<int>.Count))
+      {
+        int blocks = sp.Length / Vector<int>.Count;
+        for (int i = 0; i < blocks; i ++)
+        {
+          Vector<int> v = new Vector<int>(sp.Slice(i * Vector<int>.Count));
+          // ...
+        }              
+      }
+    }
+    ```
+
+    expands to
+
+    ```C#
+    void unsafe foo(Span<int> sp)
+    {
+      if (sp.Length >= Vector512<int>.Count)
+      {
+        int blocks = sp.Length / Vector512<int>.Count;
+        for (int i = 0; i < blocks; i ++)
+        {
+          Vector512<int> v = new Vector512<int>(sp.Slice(i * Vector512<int>.Count));
+          // ...
+        }
+      }
+      else if (sp.Length >= Vector256<int>.Count)
+      {
+        int blocks = sp.Length / Vector256<int>.Count;
+        for (int i = 0; i < blocks; i ++)
+        {
+          Vector256<int> v = new Vector256<int>(sp.Slice(i * Vector256<int>.Count));
+          // ...
+        }
+      }
+      else if (sp.Length >= Vector128<int>.Count)
+      {
+        int blocks = sp.Length / Vector128<int>.Count;
+        for (int i = 0; i < blocks; i ++)
+        {
+          Vector128<int> v = new Vector128<int>(sp.Slice(i * Vector128<int>.Count));
+          // ...
+        }
+      }
+      else
+      {
+        int blocks = sp.Length / Vector1<int>.Count;
+        for (int i = 0; i < blocks; i ++)
+        {
+          Vector1<int> v = new Vector1<int>(sp.Slice(i * Vector1<int>.Count));
+          // ...
+        }
+      }
+    }
+    ```
+
+- If the JIT encounters a `Vectorize.If` outside of an `if` conditional, the JIT will throw an `OperationNotSupportedException`.
+
+- If the JIT encounters `Vectorize.If` with `else if` or `else` arms, the JIT will throw an `OperationNotSupportedException`.
+
+- Any fixed length VectorXX will not be adjusted in any way if present in a `Vectorize.If` block.
+
+- Nested `Vectorize.If` is allowed. Expansion begins at the most nested block.
+
+#### 2. PGO Codegen from `Vector<T>` and `#[Vectorize]`
+
+In order to perform profile guided optimization for methods annotated with `#[Vectorize]`, the JIT must detect which method parameters to sample and what thresholds should trigger recompilation based on those sample points.
+##### Detecting Instrumentation Points and Thresholds
+
+The presence of a `#[Vectorize]` attribute instructs the JIT to perform a dependence analysis upon first encountering the method to determine what method parameters to sample.
```

Comment:
A of probing scheme that requires data flow is going to be a challenge.
The current implementation of PGO instruments at Tier0 where there is no ability to do any sort of data flow. 

As @EgorBo noted above it might be more feasible to do this sort of analysis either by hand at the source level or within Roslyn and express the results as intrinsic calls.

Reply:
Agreed. I have commented below.